### PR TITLE
Add SynthWordGenerator to text reco training scripts

### PR DESCRIPTION
**Pull request by felixdittrich92**
After some feedback now this PR integrates only the WordGenerator into the training scripts :smile: 

Any feedback is welcome :smiley:

### CODE REVIEW

File: doctr/datasets/generator/base.py [hunk 1]
```diff
@@ -105,9 +105,9 @@ class _WordGenerator(AbstractDataset):
     def __init__(
         self,
         vocab: str,
-        min_chars: int,
-        max_chars: int,
         num_samples: int,
+        min_chars: int = 1,
+        max_chars: int = 17,  # good match for 32x128 images with different fonts
```
Comment (fg-mindee) [inline doctr/datasets/generator/base.py:110]: The fact that our training script are used for specific shapes should not impact defaults on data generation pipelines I'd argue. I'm not sure we really need the default values here (making the text recognition script use this word generator with default values would be a better option in my opinion)

File: doctr/datasets/generator/pytorch.py [hunk 2]
```diff
@@ -51,4 +54,15 @@ class WordGenerator(_WordGenerator):
...
+        if self.img_transforms is None:
+            img_transforms = Compose([
+                T.RandomApply(RandomRotation(
+                    [-6, 6], interpolation=InterpolationMode.BILINEAR, expand=True), .2),
+                T.Resize((32, 128), preserve_aspect_ratio=True),
+                T.RandomApply(T.ColorInversion(min_val=1.0), 1.0),
+                T.RandomApply(GaussianBlur(
+                    kernel_size=3, sigma=(0.3, 2.0)), .2),
+            ])
+            setattr(self, 'img_transforms', img_transforms)
```
Comment (fg-mindee) [inline doctr/datasets/generator/pytorch.py:68]: A dataset should not have any transforms by default. This is typically the type of operations to define in the training script :+1:

File: doctr/datasets/vocabs.py [hunk 3]
```diff
@@ -28,6 +28,6 @@
 VOCABS['french'] = VOCABS['english'] + 'àâéèêëîïôùûüçÀÂÉÈÊËÎÏÔÙÛÜÇ'
 VOCABS['portuguese'] = VOCABS['english'] + 'áàâãéêíïóôõúüçÁÀÂÃÉÊÍÏÓÔÕÚÜÇ'
 VOCABS['spanish'] = VOCABS['english'] + 'áéíóúüñÁÉÍÓÚÜÑ' + '¡¿'
-VOCABS['german'] = VOCABS['english'] + 'äöüßÄÖÜẞ'
+VOCABS['german'] = VOCABS['english'] + 'äöüßÄÖÜ'
```
Comment (fg-mindee) [inline doctr/datasets/vocabs.py:31]: Oh good catch here :+1:

**Review by fg-mindee — CHANGES_REQUESTED** [review]
Thanks a lot Felix!

Would you mind just making this PR about fixing the german vocab please?
And about your suggestion to replace the character by a double S, I think it's a dangerous idea because a vocab element = a theoretical character = a computer character so far. With this, we'd be breaking this rule.

**Review by fg-mindee — CHANGES_REQUESTED** [review]
Thanks! For now, it would be good to have a similar transformation set as other recognition tasks (cf. my comments). Let me know what you think :ok_hand:

File: references/recognition/train_tensorflow.py [hunk 4]
```diff
@@ -137,13 +138,34 @@ def main(args):
...
+        )
+    else:
+        # Load synthetic data generator
+        val_set = WordGenerator(
+            vocab=VOCABS[args.vocab],
+            min_chars=1,
+            max_chars=17,
+            num_samples=int(args.num_synth_samples * 0.2),
+            font_family=[os.path.join(args.fonts_folder, f)
+                         for f in os.listdir(args.fonts_folder) if f.endswith('.ttf')],
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply references/recognition/train_tensorflow.py]: :+1:

File: references/recognition/train_tensorflow.py [hunk 5]
```diff
@@ -137,13 +138,34 @@ def main(args):
...
+            img_transforms=T.Compose([
+                T.RandomApply(T.LambdaTransformation(
+                    lambda x: rotated_img_tensor(x, np.random.choice([-6.0, 6.0]), expand=True)), .2),
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply references/recognition/train_tensorflow.py]: @fg-mindee 
for the transform module i agree !
But i am sure that a bit blurring and slight rotation (current only pytorch same in classification) will improve the performance (take a look at VITSTR). 
To ensure this i will do some tests (hopefully end of the week if i get my main laptop back :sweat_smile: ). 
Are you fine with this ? :smiley:

File: references/recognition/train_pytorch.py [hunk 6]
```diff
@@ -237,24 +257,47 @@ def main(args):
...
+                train_set.merge_dataset(RecognitionDataset(
+                    subfolder.joinpath('images'), subfolder.joinpath('labels.json')))
+        with open(parts[0].joinpath('labels.json'), 'rb') as f:
+            train_hash = hashlib.sha256(f.read()).hexdigest()
+    else:
+        # Load synthetic data generator
+        train_set = WordGenerator(
+            vocab=vocab,
+            min_chars=1,
+            max_chars=17,
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:287]: Why 17?
If you check the max_length over here https://github.com/mindee/doctr/blob/main/doctr/models/recognition/sar/pytorch.py#L157
we typically are around 30 :sweat_smile: So I'd suggest something higher

The min & max number of chars could be script args btw

**Review by fg-mindee — CHANGES_REQUESTED** [review]
Thanks for the edits! Still a few things to address and we'll be good to go!

File: references/recognition/train_pytorch.py [hunk 7]
```diff
@@ -237,24 +257,47 @@ def main(args):
...
+            font_family=fonts,
+            img_transforms=Compose([
+                T.RandomApply(RandomRotation(
+                    [-6, 6], interpolation=InterpolationMode.BILINEAR, expand=True), .2),
+                T.Resize((args.input_size, 4 * args.input_size), preserve_aspect_ratio=True),
+                # Ensure we have a 90% split of white-background images
+                T.RandomApply(T.ColorInversion(min_val=1.0), 0.9),
+                T.RandomApply(GaussianBlur(
+                    kernel_size=3, sigma=(0.3, 2.0)), .2),
+                ColorJitter(brightness=0.3, contrast=0.3, saturation=0.3, hue=0.02),
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply references/recognition/train_pytorch.py:298]: @fg-mindee 
ok but the color inversion is a must have else we will train on images with black background :sweat_smile:

File: references/recognition/train_pytorch.py (hunk 6 above)
Comment (felixdittrich92) [reply references/recognition/train_pytorch.py:287]: mh yes but this will lead with some fonts that the chars are printed outside of the image :confused:

File: references/recognition/train_pytorch.py [hunk 8]
```diff
@@ -267,8 +310,6 @@ def main(args):
     )
     print(f"Train set loaded in {time.time() - st:.4}s ({len(train_set)} samples in "
           f"{len(train_loader)} batches)")
-    with open(parts[0].joinpath('labels.json'), 'rb') as f:
-        train_hash = hashlib.sha256(f.read()).hexdigest()
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply references/recognition/train_pytorch.py:271]: it was :sweat_smile:  
I've put it at the beginning now, so it's a bit clearer :smile:

File: references/recognition/train_pytorch.py [hunk 9]
```diff
@@ -179,13 +179,36 @@ def main(args):
...
+    fonts = args.font.split(",")
+
     # Load val data generator
     st = time.time()
-    val_set = RecognitionDataset(
-        img_folder=os.path.join(args.val_path, 'images'),
-        labels_path=os.path.join(args.val_path, 'labels.json'),
-        img_transforms=T.Resize((args.input_size, 4 * args.input_size), preserve_aspect_ratio=True),
-    )
+    if args.val_path is not None:
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:187]: would you mind reversing the order and starting with the other option by doing:
```python
if isinstance(args.val_path, str):
```
would be a bit safer

File: references/recognition/train_pytorch.py [hunk 10]
```diff
@@ -179,13 +179,36 @@ def main(args):
...
+        val_set = WordGenerator(
+            vocab=vocab,
+            min_chars=args.min_chars,
+            max_chars=args.max_chars,
+            num_samples=args.val_samples * len(vocab),
+            font_family=fonts,
+            img_transforms=Compose([
+                T.Resize((args.input_size, 4 * args.input_size), preserve_aspect_ratio=True),
+                # Ensure we have a 90% split of white-background images
+                T.RandomApply(T.ColorInversion(min_val=1.0), 0.9),
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:208]: Please remove `min_val=1.0`, otherwise we only get grayscale. Without it, it allows generation of multiple contrasted colors

File: references/recognition/train_pytorch.py [hunk 11]
```diff
@@ -237,24 +258,45 @@ def main(args):
...
-            T.Resize((args.input_size, 4 * args.input_size), preserve_aspect_ratio=True),
-            # Augmentations
-            T.RandomApply(T.ColorInversion(), .1),
-            ColorJitter(brightness=0.3, contrast=0.3, saturation=0.3, hue=0.02),
-        ]),
-    )
-    if len(parts) > 1:
-        for subfolder in parts[1:]:
-            train_set.merge_dataset(RecognitionDataset(subfolder.joinpath('images'), subfolder.joinpath('labels.json')))
+    if args.train_path is not None:
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:261]: same here

File: references/recognition/train_pytorch.py [hunk 12]
```diff
@@ -237,24 +258,45 @@ def main(args):
...
+        train_set = WordGenerator(
+            vocab=vocab,
+            min_chars=args.min_chars,
+            max_chars=args.max_chars,
+            num_samples=args.train_samples * len(vocab),
+            font_family=fonts,
+            img_transforms=Compose([
+                T.Resize((args.input_size, 4 * args.input_size), preserve_aspect_ratio=True),
+                # Ensure we have a 90% split of white-background images
+                T.RandomApply(T.ColorInversion(min_val=1.0), 0.9),
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:296]: same here

File: references/recognition/train_pytorch.py [hunk 13]
```diff
@@ -348,9 +388,29 @@ def parse_args():
...
+        help='Multiplied by the vocab length gets you the number of synthetic validation samples that will be used.'
+    )
+    parser.add_argument(
+        '--font',
+        type=str,
+        default="FreeMono.ttf,FreeSans.ttf,FreeSerif.ttf",
+        help='Font family to be used'
+    )
+    parser.add_argument('--min-chars', type=int, default=1, help='Minimum number of characters per synthetic sample')
+    parser.add_argument('--max-chars', type=int, default=30, help='Maximum number of characters per synthetic sample')
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:413]: My bad, I experimented and for our default (32, 128) size, the max size that yields readable results is 12 characters.
Past that, the resizing/padding scheme makes it very hazardous to trust a human to read it

**Review by fg-mindee — CHANGES_REQUESTED** [review]
Thanks Felix!

Final adjustments to make (my comments also apply to the tensorflow script) :)

**Review by fg-mindee — APPROVED** [review]
Thanks Felix :pray:
